File: src/client/Domain/Messaging/Message.cs
using Chat.Domain.Instances;

namespace Chat.Domain.Messaging;

public enum MessageKind { Text, System, Encrypted }
public sealed record Attachment(Guid Id, string FileName, string MimeType, long Size,
    Uri DownloadUrl, Uri? ThumbnailUrl);
public sealed record Embed(string? Title, string? Description, Uri? Url);
public sealed record Reaction(string Emoji, int Count, bool ReactedByAccount);
public sealed record EncryptedPayload(string Algorithm, string KeyId, string Ciphertext);
public sealed record Message(EntityKey Key, Guid ChannelId, Guid AuthorId, MessageKind Kind,
    string? Content, DateTimeOffset CreatedAt, DateTimeOffset? EditedAt, Guid? ReplyTo,
    IReadOnlyList<Guid> Mentions, IReadOnlyList<Attachment> Attachments,
    IReadOnlyList<Embed> Embeds, IReadOnlyList<Reaction> Reactions, EncryptedPayload? Encrypted);

File: src/server/domain/src/message.rs
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct Attachment {
    pub id: Uuid,
    pub file_name: String,
    pub mime_type: String,
    pub size: u64,
    pub object_key: Uuid,
    pub thumbnail_key: Option<Uuid>,
}

#[derive(Debug, Clone)]
pub struct Message {
    pub id: Uuid,
    pub channel_id: Uuid,
    pub author_id: Uuid,
    pub kind: String,
    pub content: Option<String>,
    pub created_at: String,
    pub edited_at: Option<String>,
    pub reply_to: Option<Uuid>,
    pub attachments: Vec<Attachment>,
}
